### ISSUE TITLE
CrossAssemblySupportingAndNotSupportingCovariantReturns test throws if `GetEmitDiagnostics` call added

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#60060]
Assertion is triggered in `SynthesizedRecordEquals.GenerateMethodBody` because of an unexpected use-site diagnostic on conversion.

```
 Message: 
System.InvalidOperationException : Assertion failed

  Stack Trace: 
ThrowingTraceListener.Fail(String message, String detailMessage) line 26
TraceInternal.Fail(String message, String detailMessage)
TraceProvider.Fail(String message, String detailMessage)
Debug.Fail(String message, String detailMessage)
Debug.Assert(Boolean condition)
SyntheticBoundNodeFactory.Convert(TypeSymbol type, BoundExpression arg) line 1296
SynthesizedRecordEquals.GenerateMethodBody(TypeCompilationState compilationState, BindingDiagnosticBag diagnostics) line 129
MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) line 966
MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) line 525
<>c__DisplayClass25_0.<CompileNamedTypeAsync>b__0() line 419
<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() line 139
Task.InnerInvoke()
<.cctor>b__272_0(Object obj)
ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
MethodCompiler.WaitForWorkers() line 330
MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, Boolean emitMethodBodies, BindingDiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) line 165
CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) line 3227
Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, RebuildData rebuildData, CompilationTestData testData, CancellationToken cancellationToken) line 2837
Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, RebuildData rebuildData, CancellationToken cancellationToken) line 2778
Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, CancellationToken cancellationToken) line 2666
DiagnosticExtensions.GetEmitDiagnostics[TCompilation](TCompilation c, EmitOptions options, IEnumerable`1 manifestResources) line 360
DiagnosticExtensions.GetEmitDiagnostics[TCompilation](TCompilation c) line 372
DiagnosticExtensions.VerifyDiagnostics[TCompilation](TCompilation c, DiagnosticDescription[] expected) line 108
RecordTests.CrossAssemblySupportingAndNotSupportingCovariantReturns(Boolean useCompilationReference) line 30356
```

Relates to https://github.com/dotnet/roslyn/pull/60052

**Comment by jcouv** [comment]
Assigned to @AlekseyTs with milestone Compiler.Next. Feel free to push out to Backlog based on your judgement.
